In [1]:
from data_set import LeucorrheaDataset
import numpy as np
import time

In [2]:
data_set = LeucorrheaDataset(train=False)

In [3]:
ratios =np.array([0.54, 0.88, 1.17, 1.63, 2.3])
anchor_sizes = np.array([50, 96, 184, 273, 395])

classes =np.zeros([5,4])

num_each_class = np.zeros([7])

length = len(data_set)
for epoch in range(100):
    
    classes = classes*0
    loss = 0
    size_loss = 0
    start_time = time.time()
    for i in range(length):
        img,target = data_set[i]
        boxes = target['boxes'].numpy()
        for j in range(len(boxes)):
            box = boxes[j]
            width = box[2]-box[0]
            hight = box[3]-box[1]
            
            ratio = hight/width     
            distance = np.square(ratios-ratio)
            index = np.argmin(distance)
            loss+=distance[index]
            cp = classes[index]
            cp[0]+=ratio
            cp[1]+=1
            
            size = np.sqrt(width*hight)
            size_distance = np.square(anchor_sizes-size)
            arg_size = np.argmin(size_distance)
            size_loss +=size_distance[arg_size]
            size_cp = classes[arg_size]
            size_cp[2]+=size
            size_cp[3]+=1
        if epoch==0:
            label = target['labels']
            num_each_class[label-1]+=1
    time_elapsed = time.time()-start_time
    ratios = classes[:,0]/(classes[:,1]+1e-10)
    anchor_sizes = classes[:,2]/(classes[:,3]+1e-10)
   
    print('epoch:{}/100    ratio_loss:{:.3f}    anchor_loss:{}  time:{}'.format(epoch,loss,size_loss,time_elapsed))
    print('ratios:{}'.format(ratios))
    print('anchor_sizes:{}'.format(anchor_sizes))
    print('---------------------------------------------------------------')
print('num_each_class:{}'.format(num_each_class))
    

epoch:0/100    ratio_loss:4289.449    anchor_loss:175319264.29242092  time:627.7572135925293
ratios:[0.55105884 0.88807559 1.16256317 1.61888194 2.43895655]
anchor_sizes:[ 51.86441314  93.78506496 194.51138623 286.04885267 408.88816805]
---------------------------------------------------------------
epoch:1/100    ratio_loss:4073.632    anchor_loss:152450663.8594802  time:604.7561213970184
ratios:[0.55805933 0.89175001 1.16000939 1.62925204 2.51193083]
anchor_sizes:[ 51.80593182  93.83189201 202.89242983 297.37045636 417.46600224]
---------------------------------------------------------------
epoch:2/100    ratio_loss:4013.135    anchor_loss:139725498.29615176  time:605.4471027851105
ratios:[0.56182537 0.89385545 1.16102277 1.64219432 2.55748762]
anchor_sizes:[ 51.80593182  93.94698237 210.2019967  306.29184497 424.18252134]
---------------------------------------------------------------
epoch:3/100    ratio_loss:3984.988    anchor_loss:131549605.41703011  time:605.9691565036774
ratio

epoch:28/100    ratio_loss:3896.575    anchor_loss:114544961.09041882  time:596.5318021774292
ratios:[0.58194959 0.92665354 1.22257827 1.76515294 2.77173788]
anchor_sizes:[ 52.00721664  95.1866716  245.21359505 347.58578711 457.26316636]
---------------------------------------------------------------
epoch:29/100    ratio_loss:3896.271    anchor_loss:114543338.50057347  time:599.4029190540314
ratios:[0.58229005 0.9272125  1.22353903 1.76627514 2.77343996]
anchor_sizes:[ 52.00721664  95.1884458  245.30357695 347.69107512 457.33708337]
---------------------------------------------------------------
epoch:30/100    ratio_loss:3896.063    anchor_loss:114542132.83589123  time:603.8506731987
ratios:[0.58294754 0.92788925 1.22428308 1.76690782 2.77471898]
anchor_sizes:[ 52.00721664  95.19377279 245.39321701 347.80226564 457.44007752]
---------------------------------------------------------------
epoch:31/100    ratio_loss:3895.884    anchor_loss:114540779.87529813  time:600.6364243030548
rat

epoch:56/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:600.1857712268829
ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
epoch:57/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:601.9446179866791
ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
epoch:58/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:598.9874339103699
ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
epoch:59/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:601.184160232544
r

epoch:84/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:599.2507412433624
ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
epoch:85/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:599.9828841686249
ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
epoch:86/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:597.4408195018768
ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
epoch:87/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:596.9920291900635


- epoch:99/100    ratio_loss:3895.637    anchor_loss:114538157.51998879  time:598.2679386138916
- ratios:[0.58415315 0.92982373 1.22694066 1.76789341 2.77628675]
- anchor_sizes:[ 52.01049868  95.21385875 245.83876798 348.32761324 457.85658187]
---------------------------------------------------------------
- num_each_class:[12099. 16785.  9101.  2494.  1337.   232.   295.]

In [14]:
weights = 1/(num_each_class/np.sum(num_each_class)) 
w = 16785/num_each_class
print(w)

[ 1.38730474  1.          1.84430282  6.73015237 12.55422588 72.34913793
 56.89830508]


AttributeError: 'tuple' object has no attribute 'shape'